### ISSUE TITLE
Brian2 back-end development for PyNN

### ISSUE THREAD

**Issue by appukuttan-shailesh** [NeuralEnsemble/PyNN#632]
We finally have a very early stage prototype for Brian2 back-end for PyNN. See **brian2_dev** branch. The dev work basically involved using the existing Brian back-end and updating the source files as necessary to align with changes introduced in Brian2.

Some of the currently implemented features:
- Various cell types
- Current injection into single cell
- All the electrode types
- Recording cell and electrode parameters

Source Code:
```
import sys
from pyNN.utility import get_script_args

simulator_name = get_script_args(1)[0]
exec("from pyNN.%s import *" % simulator_name)

# setup(timestep=0.01, min_delay=0.1)
setup(min_delay=0.1)

cells = Population(1, IF_curr_exp(v_rest = -65.0, v_thresh=-55.0, tau_refrac=5.0))#, i_offset=-1.0*amp))

dc_source = DCSource(amplitude=0.5, start=25, stop=50)
ac_source = ACSource(start=75, stop=125, amplitude=0.5, offset=0.25, frequency=100.0, phase=0.0)
noisy_source = NoisyCurrentSource(mean=0.5, stdev=0.05, start=150, stop=175, dt=1.0)
step_source = StepCurrentSource(times=[200, 225, 250], amplitudes=[0.4, 0.6, 0.2])

cells[0].inject(dc_source)
cells[0].inject(ac_source)
cells[0].inject(noisy_source)
cells[0].inject(step_source)

dc_source.record()
ac_source.record()
noisy_source.record()
step_source.record()

cells.record(['v'])

run(275.0)

if '--plot-figure' in sys.argv:
    import matplotlib.pyplot as plt
    plt.ion()

    plt.figure(figsize=(8,8))

    plt.subplot(2,1,1)
    vm = cells.get_data().segments[0].filter(name="v")[0]
    plt.plot(vm.times, vm, 'r')
    plt.xlabel("time (ms)")
    plt.ylabel("Vm (mV)")

    plt.subplot(2,1,2)
    i_t_dc, i_amp_dc = dc_source._get_data()
    i_t_ac, i_amp_ac = ac_source._get_data()
    i_t_noisy, i_amp_noisy = noisy_source._get_data()
    i_t_step, i_amp_step = step_source._get_data()
    plt.plot(i_t_dc, i_amp_dc, 'k')
    plt.plot(i_t_ac, i_amp_ac, 'k')
    plt.plot(i_t_noisy, i_amp_noisy, 'k')
    plt.plot(i_t_step, i_amp_step, 'k')
    plt.xlabel("time (ms)")
    plt.ylabel("current (nA)")

    plt.suptitle(simulator_name, fontsize=15)
    plt.show(block=True)  # SA: changed
end()

```
![image](https://user-images.githubusercontent.com/24866517/56592216-833bd580-65ea-11e9-8e45-7846bfe32da0.png)

Known issues:
1) Handling current injection on population of size >1
2) Handling current injection on multiple populations
3) ~~Errors described in #630 and #631~~
4) Unable to print individual neuron parameters (e.g. `v_thresh`)
5) `i_offset` having to be specified with units

(please add to this list)

Things to be checked/implemented/tested:
1) Projections
2) Synapes

(please add to this list)

If you wish to work on an issue/pending item listed above, create a new ticket for each item and provide details.

**Note:** Use a version of LazyArray that includes the following change:
> https://github.com/NeuralEnsemble/lazyarray/commit/38850fc9077a666e8feb225272d6fb4745642c7e

**Comment by appukuttan-shailesh** [comment]
@sanjayankur31 , @laurentperrinet : tagging you both since you had expressed an interest towards this work.

**Comment by appukuttan-shailesh** [comment]
Issues #630 and #631 resolved for Brian2.

![image](https://user-images.githubusercontent.com/24866517/57143067-55ab1500-6dbe-11e9-83e2-7a7d9e8e0375.png)

**Comment by laurentperrinet** [comment]
using that code 

```
import pyNN.brian2 as sim
sim.setup(min_delay=0.1)
cells = sim.Population(1, sim.IF_curr_exp(v_rest = -65.0, v_thresh=-55.0, tau_refrac=5.0))#, i_offset=-1.0*amp))
noisy_source = sim.NoisyCurrentSource(mean=0.5, stdev=0.05, start=150, stop=175, dt=1.0)
```

I get the error

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/usr/local/lib/python3.7/site-packages/pyNN/standardmodels/__init__.py in __getattr__(self, name)
    171         try:
--> 172             val = self.__getattribute__(name)
    173         except AttributeError:

AttributeError: 'NoisyCurrentSource' object has no attribute 'running'

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
<ipython-input-25-65b18109ae8a> in <module>
      2 sim.setup(min_delay=0.1)
      3 cells = sim.Population(1, sim.IF_curr_exp(v_rest = -65.0, v_thresh=-55.0, tau_refrac=5.0))#, i_offset=-1.0*amp))
----> 4 noisy_source = sim.NoisyCurrentSource(mean=0.5, stdev=0.05, start=150, stop=175, dt=1.0)

/usr/local/lib/python3.7/site-packages/pyNN/brian2/standardmodels/electrodes.py in __init__(self, **parameters)
    231 
    232     def __init__(self, **parameters):
--> 233         Brian2CurrentSource.__init__(self, **parameters)
    234         self._generate()
    235 

/usr/local/lib/python3.7/site-packages/pyNN/brian2/standardmodels/electrodes.py in __init__(self, **parameters)
     41         parameter_space.update(**parameters)
     42         parameter_space = self.translate(parameter_space)
---> 43         self.set_native_parameters(parameter_space)
     44 
     45     def _check_step_times(self, times, amplitudes, resolution):

/usr/local/lib/python3.7/site-packages/pyNN/brian2/standardmodels/electrodes.py in set_native_parameters(self, parameters)
     78                 value = value.value
     79             object.__setattr__(self, name, value)
---> 80         self._reset()
     81 
     82     def _reset(self):

/usr/local/lib/python3.7/site-packages/pyNN/brian2/standardmodels/electrodes.py in _reset(self)
     82     def _reset(self):
     83         # self.i reset to 0 only at the start of a new run; not for continuation of existing runs
---> 84         if not hasattr(self, 'running') or self.running == False:
     85             self.i = 0
     86             self.running = True

/usr/local/lib/python3.7/site-packages/pyNN/standardmodels/__init__.py in __getattr__(self, name)
    173         except AttributeError:
    174             try:
--> 175                 val = self.get_parameters()[name]
    176             except KeyError:
    177                 raise errors.NonExistentParameterError(name,

/usr/local/lib/python3.7/site-packages/pyNN/standardmodels/__init__.py in get_parameters(self)
    206     def get_parameters(self):
    207         """Return a dict of all current source parameters."""
--> 208         parameters = self.get_native_parameters()
    209         parameters = self.reverse_translate(parameters)
    210         return parameters

/usr/local/lib/python3.7/site-packages/pyNN/standardmodels/__init__.py in get_native_parameters(self)
    214 
    215     def get_native_parameters(self):
--> 216         raise NotImplementedError
    217 
    218     def _round_timestamp(self, value, resolution):

NotImplementedError: 
```

**Comment by apdavison** [comment]
closing this as we released Brian 2 support with v0.9.6. Please open individual tickets for any remaining bugs you may find.
